{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3d0f8ec7-ac53-4686-9dd0-112f9e2c6888",
   "metadata": {},
   "source": [
    "# Create sample map tsv files\n",
    "\n",
    "- generate four sex-specific TSVs that point from each sample to where the following files (generated by Step 3) exist:\n",
    "    - `chrX_PAR1_sample_map.tsv`\n",
    "        - For XY samples, map `sample_id` to `XY_X_PAR_hcVCF_gz` output from Step 3\n",
    "        - For XX samples, map `sample_id` to `XX_X_hcVCF_gz` output from Step 3\n",
    "    - `chrX_PAR2_sample_map.tsv`\n",
    "        - Same as above\n",
    "    - `chrX_non_PAR_sample_map.tsv`\n",
    "        - For XY samples, map `sample_id` to `XY_X_non_PAR_hcVCF_gz` output from Step 3\n",
    "        - For XX samples, map `sample_id` to `XX_X_hcVCF_gz` output from Step 3\n",
    "    - `chrY_sample_map.tsv`\n",
    "        - For XY samples, map `sample_id` to `XY_Y_nonPAR_hcVCF_gz` output from Step 3\n",
    "        - Do not include XX samples\n",
    "- generate remainder mapping files for autosomomal chromosomes\n",
    "    - `chr<#>_sample_map.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1e1a4a5a-a131-465b-9c29-96f42174a34e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5d911a39-36ec-45f9-a84c-05a84b440bb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_sample_autosomal_chr(submission_tsv, autosomal_chr_map_list):\n",
    "    chrms = list(range(22))\n",
    "    \n",
    "    submission_id = submission_tsv.strip('.tsv').split('/')[-1]\n",
    "    submission_tsv_df = pd.read_csv(submission_tsv, sep='\\t')\n",
    "    all_samples = list(submission_tsv_df['asText'])\n",
    "\n",
    "    for sample in all_samples:\n",
    "        sample_split = sample.strip().replace('\"','').split(' ')\n",
    "\n",
    "        sample_name = sample_split[3]\n",
    "        workflow_id = sample_split[4]\n",
    "\n",
    "        # get gsutil urls\n",
    "        bashCommand = \"gsutil ls gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/submissions/\" + submission_id + \"/haplotype_calling_chrom/\" + workflow_id + \"/**/*vcf*\"\n",
    "        process = subprocess.Popen(bashCommand.split(), stdout=subprocess.PIPE)\n",
    "        outputs, error = process.communicate()\n",
    "        outputs = outputs.decode(\"utf-8\").replace('\\n',' ').strip().split()\n",
    "\n",
    "        # assign autosomal chromosome maps\n",
    "        candidate_chr_idx_sample_map_urls = [[] for _ in chrms]   # make empty list of lists\n",
    "\n",
    "        for chrm_idx in chrms:            \n",
    "            for output in outputs:\n",
    "                if output.endswith( '.chr' + str(chrm_idx+1) + '.hc.vcf.gz'):\n",
    "                    candidate_chr_idx_sample_map_urls[chrm_idx].append(output)\n",
    "\n",
    "            # select url from last attempt and map sample_name to file\n",
    "            autosomal_chr_map_list[chrm_idx][sample_name] = sorted(candidate_chr_idx_sample_map_urls[chrm_idx])[-1]\n",
    "\n",
    "    return autosomal_chr_map_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b55ed42-83f2-4859-8f87-481bdde865cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_sample_sex_chr(submission_tsv,\n",
    "        chrX_PAR1_sample_map,\n",
    "        chrX_PAR2_sample_map,\n",
    "        chrX_non_PAR_sample_map,\n",
    "        chrY_sample_map):\n",
    "\n",
    "    submission_id = submission_tsv.strip('.tsv').split('/')[-1]\n",
    "\n",
    "    submission_tsv_df = pd.read_csv(submission_tsv, sep='\\t')\n",
    "    all_samples = list(submission_tsv_df['asText'])\n",
    "\n",
    "    for sample in all_samples:\n",
    "        sample_split = sample.strip().replace('\"','').split(' ')\n",
    "\n",
    "        sex = sample_split[-1]\n",
    "        sample_name = sample_split[3]\n",
    "        workflow_id = sample_split[4]  #  d6795d02-f59e-48c0-89bf-51abc4d9160b\n",
    "\n",
    "        # get gsutil urls\n",
    "        bashCommand = \"gsutil ls gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/submissions/\" + submission_id + \"/haplotype_calling_chrom/\" + workflow_id + \"/**/*vcf*\"\n",
    "        process = subprocess.Popen(bashCommand.split(), stdout=subprocess.PIPE)\n",
    "        outputs, error = process.communicate()\n",
    "        outputs = outputs.decode(\"utf-8\").replace('\\n',' ').strip().split()\n",
    "\n",
    "        # assign to proper files\n",
    "        if sex == 'male':\n",
    "            candidate_chrX_PAR_sample_map_urls = []\n",
    "            candidate_chrX_non_PAR_sample_map_urls = []\n",
    "            candidate_chrY_sample_map_urls = []\n",
    "\n",
    "            for output in outputs:\n",
    "                # add to: chrX_PAR1_sample_map.tsv and chrX_PAR2_sample_map.tsv\n",
    "                #   For XY samples, map sample_id to XY_X_PAR_hcVCF_gz output from Step 3\n",
    "                if output.endswith('.chrX_PAR.hc.vcf.gz'):\n",
    "                    candidate_chrX_PAR_sample_map_urls.append(output)\n",
    "\n",
    "                # add to: chrX_non_PAR_sample_map.tsv\n",
    "                #   For XY samples, map sample_id to XY_X_non_PAR_hcVCF_gz output from Step 3\n",
    "                if output.endswith('.chrX_non_PAR.hc.vcf.gz'):\n",
    "                    candidate_chrX_non_PAR_sample_map_urls.append(output)\n",
    "\n",
    "                # add to: chrY_sample_map.tsv\n",
    "                #   For XY samples, map sample_id to XY_Y_nonPAR_hcVCF_gz output from Step 3\n",
    "                if output.endswith('.chrY.hc.vcf.gz'):\n",
    "                    candidate_chrY_sample_map_urls.append(output)\n",
    "\n",
    "            # select url from last attempt\n",
    "            chrX_PAR_sample_map_url = sorted(candidate_chrX_PAR_sample_map_urls)[-1]\n",
    "            chrX_non_PAR_sample_map_url = sorted(candidate_chrX_non_PAR_sample_map_urls)[-1]\n",
    "            chrY_sample_map_url = sorted(candidate_chrY_sample_map_urls)[-1]\n",
    "\n",
    "            # map sample_name to file\n",
    "            chrX_PAR1_sample_map[sample_name] = chrX_PAR_sample_map_url\n",
    "            chrX_PAR2_sample_map[sample_name] = chrX_PAR_sample_map_url\n",
    "            chrX_non_PAR_sample_map[sample_name] = chrX_non_PAR_sample_map_url\n",
    "            chrY_sample_map[sample_name] = chrY_sample_map_url\n",
    "\n",
    "        elif sex == 'female':\n",
    "            candidate_chrX_PAR_sample_map_urls = []\n",
    "            for output in outputs:\n",
    "                if output.endswith('.chrX.hc.vcf.gz'):\n",
    "                    candidate_chrX_PAR_sample_map_urls.append(output)\n",
    "\n",
    "            # select url from last attempt\n",
    "            chrX_PAR_sample_map_url = sorted(candidate_chrX_PAR_sample_map_urls)[-1]\n",
    "\n",
    "            # map sample_name to file\n",
    "            chrX_PAR1_sample_map[sample_name] = chrX_PAR_sample_map_url\n",
    "            chrX_PAR2_sample_map[sample_name] = chrX_PAR_sample_map_url\n",
    "            chrX_non_PAR_sample_map[sample_name] = chrX_PAR_sample_map_url\n",
    "        else:\n",
    "            raise Exception('Unlabeled sex')\n",
    "\n",
    "    return chrX_PAR1_sample_map, chrX_PAR2_sample_map, chrX_non_PAR_sample_map, chrY_sample_map\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2bf78fc1-9910-477c-a3c8-c54753f6bf56",
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_tsv_male = '/home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/25a197b7-7803-4ebb-9550-ec8187d04cbd.tsv'\n",
    "submission_tsv_female = '/home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/d8c61c51-1023-4024-89e2-5c3856a5237c.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "72fe7258-0a8b-468c-88fb-5e10e708742e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# map autosomal chromosomes\n",
    "chrms = list(range(22))\n",
    "autosomal_chr_map_list = [{} for _ in chrms]\n",
    "\n",
    "autosomal_chr_map_list = map_sample_autosomal_chr(submission_tsv_male, autosomal_chr_map_list)\n",
    "autosomal_chr_map_list = map_sample_autosomal_chr(submission_tsv_female, autosomal_chr_map_list)\n",
    "\n",
    "chr1_sample_map, chr2_sample_map, chr3_sample_map, chr4_sample_map, chr5_sample_map, chr6_sample_map, chr7_sample_map, chr8_sample_map, chr9_sample_map, chr10_sample_map, chr11_sample_map, chr12_sample_map, chr13_sample_map, chr14_sample_map, chr15_sample_map, chr16_sample_map, chr17_sample_map, chr18_sample_map, chr19_sample_map, chr20_sample_map, chr21_sample_map, chr22_sample_map = autosomal_chr_map_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bf39aaa3-d751-4569-a234-fc2c732e3c27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# files that we want\n",
    "chr1_sample_map_df  = pd.DataFrame.from_dict(chr1_sample_map, orient='index')\n",
    "chr2_sample_map_df  = pd.DataFrame.from_dict(chr2_sample_map, orient='index')\n",
    "chr3_sample_map_df  = pd.DataFrame.from_dict(chr3_sample_map, orient='index')\n",
    "chr4_sample_map_df  = pd.DataFrame.from_dict(chr4_sample_map, orient='index')\n",
    "chr5_sample_map_df  = pd.DataFrame.from_dict(chr5_sample_map, orient='index')\n",
    "chr6_sample_map_df  = pd.DataFrame.from_dict(chr6_sample_map, orient='index')\n",
    "chr7_sample_map_df  = pd.DataFrame.from_dict(chr7_sample_map, orient='index')\n",
    "chr8_sample_map_df  = pd.DataFrame.from_dict(chr8_sample_map, orient='index')\n",
    "chr9_sample_map_df  = pd.DataFrame.from_dict(chr9_sample_map, orient='index')\n",
    "chr10_sample_map_df = pd.DataFrame.from_dict(chr10_sample_map, orient='index')\n",
    "chr11_sample_map_df = pd.DataFrame.from_dict(chr11_sample_map, orient='index')\n",
    "chr12_sample_map_df = pd.DataFrame.from_dict(chr12_sample_map, orient='index')\n",
    "chr13_sample_map_df = pd.DataFrame.from_dict(chr13_sample_map, orient='index')\n",
    "chr14_sample_map_df = pd.DataFrame.from_dict(chr14_sample_map, orient='index')\n",
    "chr15_sample_map_df = pd.DataFrame.from_dict(chr15_sample_map, orient='index')\n",
    "chr16_sample_map_df = pd.DataFrame.from_dict(chr16_sample_map, orient='index')\n",
    "chr17_sample_map_df = pd.DataFrame.from_dict(chr17_sample_map, orient='index')\n",
    "chr18_sample_map_df = pd.DataFrame.from_dict(chr18_sample_map, orient='index')\n",
    "chr19_sample_map_df = pd.DataFrame.from_dict(chr19_sample_map, orient='index')\n",
    "chr20_sample_map_df = pd.DataFrame.from_dict(chr20_sample_map, orient='index')\n",
    "chr21_sample_map_df = pd.DataFrame.from_dict(chr21_sample_map, orient='index')\n",
    "chr22_sample_map_df = pd.DataFrame.from_dict(chr22_sample_map, orient='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9318f082-5858-4b6f-95b8-9d9149519703",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write to tsv\n",
    "base_dir = '/home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/'\n",
    "\n",
    "chr1_sample_map_df.to_csv(base_dir  + 'chr1_sample_map.tsv',  sep=\"\\t\", header=False)\n",
    "chr2_sample_map_df.to_csv(base_dir  + 'chr2_sample_map.tsv',  sep=\"\\t\", header=False)\n",
    "chr3_sample_map_df.to_csv(base_dir  + 'chr3_sample_map.tsv',  sep=\"\\t\", header=False)\n",
    "chr4_sample_map_df.to_csv(base_dir  + 'chr4_sample_map.tsv',  sep=\"\\t\", header=False)\n",
    "chr5_sample_map_df.to_csv(base_dir  + 'chr5_sample_map.tsv',  sep=\"\\t\", header=False)\n",
    "chr6_sample_map_df.to_csv(base_dir  + 'chr6_sample_map.tsv',  sep=\"\\t\", header=False)\n",
    "chr7_sample_map_df.to_csv(base_dir  + 'chr7_sample_map.tsv',  sep=\"\\t\", header=False)\n",
    "chr8_sample_map_df.to_csv(base_dir  + 'chr8_sample_map.tsv',  sep=\"\\t\", header=False)\n",
    "chr9_sample_map_df.to_csv(base_dir  + 'chr9_sample_map.tsv',  sep=\"\\t\", header=False)\n",
    "chr10_sample_map_df.to_csv(base_dir + 'chr10_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr11_sample_map_df.to_csv(base_dir + 'chr11_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr12_sample_map_df.to_csv(base_dir + 'chr12_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr13_sample_map_df.to_csv(base_dir + 'chr13_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr14_sample_map_df.to_csv(base_dir + 'chr14_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr15_sample_map_df.to_csv(base_dir + 'chr15_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr16_sample_map_df.to_csv(base_dir + 'chr16_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr17_sample_map_df.to_csv(base_dir + 'chr17_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr18_sample_map_df.to_csv(base_dir + 'chr18_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr19_sample_map_df.to_csv(base_dir + 'chr19_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr20_sample_map_df.to_csv(base_dir + 'chr20_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr21_sample_map_df.to_csv(base_dir + 'chr21_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chr22_sample_map_df.to_csv(base_dir + 'chr22_sample_map.tsv', sep=\"\\t\", header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea4400ed-eb33-46c4-9036-ef97a9f421d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e603c931-9e35-4150-89a2-9594ae0f15c0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2f16b3ed-152b-4b96-a0c7-4b7e84243999",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrX_PAR1_sample_map = {}\n",
    "chrX_PAR2_sample_map = {}\n",
    "chrX_non_PAR_sample_map = {}\n",
    "chrY_sample_map = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "59a858f7-b921-49a6-8e5d-8920d2950250",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# map male samples\n",
    "chrX_PAR1_sample_map, chrX_PAR2_sample_map, chrX_non_PAR_sample_map, chrY_sample_map = map_sample_sex_chr(submission_tsv_male,\n",
    "                                                                                            chrX_PAR1_sample_map,\n",
    "                                                                                            chrX_PAR2_sample_map,\n",
    "                                                                                            chrX_non_PAR_sample_map,\n",
    "                                                                                            chrY_sample_map)\n",
    "\n",
    "# map female samples\n",
    "chrX_PAR1_sample_map, chrX_PAR2_sample_map, chrX_non_PAR_sample_map, chrY_sample_map = map_sample_sex_chr(submission_tsv_female,\n",
    "                                                                                            chrX_PAR1_sample_map,\n",
    "                                                                                            chrX_PAR2_sample_map,\n",
    "                                                                                            chrX_non_PAR_sample_map,\n",
    "                                                                                            chrY_sample_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b79769b3-45a2-45c9-b1d0-f4adaa634341",
   "metadata": {},
   "outputs": [],
   "source": [
    "# files that we want\n",
    "chrX_PAR1_sample_map_df = pd.DataFrame.from_dict(chrX_PAR1_sample_map, orient='index')\n",
    "chrX_PAR2_sample_map_df = pd.DataFrame.from_dict(chrX_PAR2_sample_map, orient='index')\n",
    "chrX_non_PAR_sample_map_df = pd.DataFrame.from_dict(chrX_non_PAR_sample_map, orient='index')\n",
    "chrY_sample_map_df = pd.DataFrame.from_dict(chrY_sample_map, orient='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "55a48ddf-d23b-4a21-add4-58ac2c56ef2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write to tsv\n",
    "base_dir = '/home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/'\n",
    "\n",
    "chrX_PAR1_sample_map_df.to_csv(base_dir + 'chrX_PAR1_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chrX_PAR2_sample_map_df.to_csv(base_dir + 'chrX_PAR2_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chrX_non_PAR_sample_map_df.to_csv(base_dir + 'chrX_non_PAR_sample_map.tsv', sep=\"\\t\", header=False)\n",
    "chrY_sample_map_df.to_csv(base_dir + 'chrY_sample_map.tsv', sep=\"\\t\", header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e00fde62-d051-44bd-a7c9-1d5f9d1e75ad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e61bfaa7-dec1-4c17-bd35-a1bb3dcd1008",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_12</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_139</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_207</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_22</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_306</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_11</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_132</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_14</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_186</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simons_data_sample_45</th>\n",
       "      <td>gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        0\n",
       "simons_data_sample_12   gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...\n",
       "simons_data_sample_139  gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...\n",
       "simons_data_sample_207  gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...\n",
       "simons_data_sample_22   gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...\n",
       "simons_data_sample_306  gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...\n",
       "simons_data_sample_11   gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...\n",
       "simons_data_sample_132  gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...\n",
       "simons_data_sample_14   gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...\n",
       "simons_data_sample_186  gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s...\n",
       "simons_data_sample_45   gs://fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70/s..."
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# view files\n",
    "chrX_PAR1_sample_map_df\n",
    "#chrX_PAR2_sample_map_df\n",
    "#chrX_non_PAR_sample_map_df\n",
    "#chrY_sample_map_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ffefe16-1067-4b42-b8cd-9e2d327fa0b4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b986c0f0-9926-4d1c-951a-fda1830aaff3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr10_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr11_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr12_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr13_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "- [4 files][  9.0 KiB/  9.0 KiB]                                                \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr14_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr15_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr16_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr17_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr18_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr19_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr1_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr20_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr21_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr22_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr2_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr3_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr4_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr5_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr6_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr7_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr8_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chr9_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chrX_non_PAR_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chrX_PAR1_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chrX_PAR2_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "Copying file:///home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps/chrY_sample_map.tsv [Content-Type=text/tab-separated-values]...\n",
      "- [26 files][ 57.5 KiB/ 57.5 KiB]                                               \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "\n",
      "Operation completed over 26 objects/57.5 KiB.                                    \n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# upload tsv to gsutils bucket\n",
    "\n",
    "# bucket\n",
    "#https://console.cloud.google.com/storage/browser/fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70\n",
    "# make new folder under 'uploads':  'sample_maps'\n",
    "\n",
    "base_dir=/home/stephen/Documents/projects/schatz_lab/rotation/t2t-sgdp/data/sample_maps\n",
    "bucket_id=fc-1b543b1d-8887-4a24-a4fd-37f06e9d9d70\n",
    "gsutil cp $base_dir/* gs://$bucket_id/uploads/sample_maps/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c720e14-7ae4-425b-948a-0d0699a2b2cf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
